### ISSUE TITLE
`--inspect-brk` runs immediately when the debugger attaches

### ISSUE THREAD

**Issue by leduyquang753** [vitest-dev/vitest#6103]
### Describe the bug

When a debugger like Chrome DevTools attaches to `vitest --inspect-brk`, the code is immediately and prematurely resumed execution.

It appears that the execution is resumed when `Runtime.runIfWaitingForDebugger` is called by the debugger. The execution is still paused when using the normal `node --inspect-brk`.

### Reproduction

Very simple source files:

`test.ts`:

```ts
export function doubleUp(n: number): number {
	if (n > 0) n = n * 2;
	return n;
}
```

`test.spec.ts`:
```ts
import {expect, test} from "vitest";
import {doubleUp} from "./test";

test("Test", () => {
	expect(doubleUp(5)).toBe(10);
});
```

Run `vitest --run --inspect-brk --no-file-parallelism` and attach a debugger such as Chrome DevTools.

### System Info

```shell
System:
  OS: Windows 11 10.0.22631
  CPU: (8) x64 Intel(R) Core(TM) i5-1035G1 CPU @ 1,00GHz
  Memory: 19,78 GB
Binaries:
  Node: 22.3.0 - C:\Program Files\NodeJS\node.EXE
  Yarn: 1.22.19 - ~\AppData\Roaming\npm\yarn.CMD
  npm: 10.8.1 - C:\Program Files\NodeJS\npm.CMD
Browsers:
  Edge: Chromium (126.0.2592.87)
  Internet Explorer: 11.0.22621.3527

vitest/2.0.2 win32-x64 node-v22.3.0
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by AriPerkkio** [comment]
I'm unable to reproduce this. Debugger stops immediately on top of the `test.spec.ts` when dev tools are attached. 

![image](https://github.com/user-attachments/assets/7d0978b0-47a9-491e-9b1a-21e966fbe55e)

**Comment by leduyquang753** [comment]
> I'm unable to reproduce this. Debugger stops immediately on top of the `test.spec.ts` when dev tools are attached. This was implemented in #5355.
> 
> ![image](https://private-user-images.githubusercontent.com/14806298/348296698-7d0978b0-47a9-491e-9b1a-21e966fbe55e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA3OTUyMDIsIm5iZiI6MTcyMDc5NDkwMiwicGF0aCI6Ii8xNDgwNjI5OC8zNDgyOTY2OTgtN2QwOTc4YjAtNDdhOS00OTFlLTliMWEtMjFlOTY2ZmJlNTVlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzEyVDE0MzUwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNkYjYyNTA1YWFiNjkyNmM0MzQ2MjU3ZmY2NGYwYzNhOTY5NjVmN2MxNjA3NzQ0MmI4MzhiYmZlNjgyMzQ5ZDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0._FMgvjslUkiA-WasIOOqGxViru4gHIXdQfOSTZpIUsA)

It might be dependent on the environment as that's macOS and I'm on Windows. Then what kind of "reproduction" do I have to do then? Screen recording?

**Comment by AriPerkkio** [comment]
The test case seems to be skipped on Windows too. Maybe there is indeed something broken.

https://github.com/vitest-dev/vitest/blob/bd83f6cadcf2f719f8b23dc0e26f83229173e596/test/cli/test/inspect.test.ts#L10

As most of us don't have access to Windows machines it would be amazing to get pull request with the fix for this. The related code is around here:

https://github.com/vitest-dev/vitest/blob/bd83f6cadcf2f719f8b23dc0e26f83229173e596/packages/vitest/src/runtime/inspector.ts#L9-L13

**Comment by leduyquang753** [comment]
Found the issue. It's that on Windows, file paths begin with the drive letter `D:` which when passed to `URL` on [line 40](https://github.com/vitest-dev/vitest/blob/bd83f6cadcf2f719f8b23dc0e26f83229173e596/packages/vitest/src/runtime/inspector.ts#L40) is mistaken as the URL's protocol. That, along with several other failing tests as I ran the test suite, show me how much Vitest actually cares about Windows.

The bigger issue now is that when Vitest is run by `inspect.test.ts`, `inspector.enabled` becomes `false` for some oddball reason that I cannot debug with Visual studio code, nor were the various `console.trace`s I sprinkle into the codebase hit. Would be great to have some assistance but I guess not.

**Comment by AriPerkkio** [comment]
> Found the issue. It's that on Windows, file paths begin with the drive letter `D:` which when passed to `URL` on [line 40](https://github.com/vitest-dev/vitest/blob/bd83f6cadcf2f719f8b23dc0e26f83229173e596/packages/vitest/src/runtime/inspector.ts#L40) is mistaken as the URL's protocol. 

Awesome, thanks for looking into this! Could you try if this minimal repro works on Windows? If not, what changes are required for it to work? https://github.com/vitest-dev/vitest/pull/5355#discussion_r1516722085


> That, along with several other failing tests as I ran the test suite [...]

We run all tests on CI on Windows too. The ones that are marked as `skipIf(isWindow)` are indeed skipped on Windows. Do you mean you removed some of those `skipIf()` and saw that the tests failed? 

> [...] shows me how much Vitest actually cares about Windows.

Do you have any suggestions how this could be improved? Always happy to extend support for all platforms. At least I thought our Windows support is very good. 🙌 

> for some oddball reason that I cannot debug with Visual studio code, nor were the various `console.traces` I sprinkled into the codebase hit. Would be great to have some assistance but I guess not.

I'm not sure I'm following here, but to me it sounds like you are seeing NodeJS issues with `worker_threads` debugging. It's simply not possible with some debuggers. Happy to help if you could write down the steps you did.

**Comment by leduyquang753** [comment]
> Could you try if this minimal repro works on Windows?

It does work and break properly.

> We run all tests on CI on Windows too. The ones that are marked as `skipIf(isWindow)` are indeed skipped on Windows. Do you mean you removed some of those `skipIf()` and saw that the tests failed?

I had not done anything with the codebase at the time. I cloned the repository, set it up and ran the test suite right away. (`test:all` doesn't exist as a script, fix your docs.) Here are the failing ones: [failing.txt](https://github.com/user-attachments/files/16200683/failing.txt).

> I'm not sure I'm following here, but to me it sounds like you are seeing NodeJS issues with `worker_threads` debugging. It's simply not possible with some debuggers. Happy to help if you could write down the steps you did.

So the `console.trace`s were after I gave up on using Visual studio code and I just manually added `console.trace()` lines into where I thought the code must have been going through like `resolveConfig` or even `Logger.printBanner`, then ran `pnpm --filter vitest build`. No trace was printed when I ran `vitest` after that, only `console.log`s I wrote into `inspector.ts` and `worker.ts` worked.

**Comment by AriPerkkio** [comment]
> Here are the failing ones

I'm also seeing these failures locally. Looks like all tests require `CI=true` flag that's passed by `pnpm test:ci`. We'll need to fix these internal tests.


> It does work and break properly.

If I change the relative filepath to full absolute one, it doesn't work on Windows anymore. :thinking:

Both of the versions below work on all platforms except Windows. After the change it works on Windows too, weird.

```diff
const filename = resolve("./target.mjs");
-const url = new URL(filename, import.meta.url).href; // Does not work on Windows
+const url = pathToFileURL(filename); // Works on all platforms

session.post("Debugger.setBreakpointByUrl", { url, lineNumber: 0 });
```

There's fix for the original issue on #6110. It also enables the test case on Windows CI which seems to pass fine now. Thanks for the root cause analysis! 

If there's anything else that doesn't work on your final project on Windows, feel free to file bug reports for those. I'm still surprised to hear that this small bug on `--inspect-brk` gives the impression that Vitest doesn't care about Windows. Hopefully #6110 changes your view once next version is released. 🙌
